// powerbi-visuals-utils-chartutils/lib/label/labelUtils.js
/*
 *  Power BI Visualizations
 *
 *  Copyright (c) Microsoft Corporation
 *  All rights reserved.
 *  MIT License
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the ""Software""), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 */
import { CssConstants } from "powerbi-visuals-utils-svgutils";
import { drawDefaultLabels as newDrawDefaultLabels, getDataLabelLayoutOptions as newGetDataLabelLayoutOptions, getNumberOfLabelsToRender as newGetNumberOfLabelsToRender } from "./newDataLabelUtils";
export const DefaultFontSizeInPt = 9;
export const horizontalLabelBackgroundPadding = 4;
export const verticalLabelBackgroundPadding = 2;
export const labelGraphicsContextClass = CssConstants.createClassAndSelector("labelGraphicsContext");
export const labelBackgroundGraphicsContextClass = CssConstants.createClassAndSelector("labelBackgroundGraphicsContext");
export function downgradeToOldLabels(labels) {
    if (!labels)
        return;
    return labels.map((label) => {
        var _a, _b, _c, _d, _e, _f, _g;
        const inheritedLabel = Object.assign({}, label);
        inheritedLabel.fontProperties = null;
        const oldLabel = inheritedLabel;
        oldLabel.fill = (_a = label === null || label === void 0 ? void 0 : label.fontProperties) === null || _a === void 0 ? void 0 : _a.color;
        oldLabel.fontSize = (_c = (_b = label === null || label === void 0 ? void 0 : label.fontProperties) === null || _b === void 0 ? void 0 : _b.size) === null || _c === void 0 ? void 0 : _c.pt;
        oldLabel.fontFamily = (_d = label === null || label === void 0 ? void 0 : label.fontProperties) === null || _d === void 0 ? void 0 : _d.family;
        oldLabel.fontWeight = (_e = label === null || label === void 0 ? void 0 : label.fontProperties) === null || _e === void 0 ? void 0 : _e.weight;
        oldLabel.fontStyle = (_f = label === null || label === void 0 ? void 0 : label.fontProperties) === null || _f === void 0 ? void 0 : _f.style;
        oldLabel.textDecoration = (_g = label === null || label === void 0 ? void 0 : label.fontProperties) === null || _g === void 0 ? void 0 : _g.decoration;
        return oldLabel;
    });
}
export function drawDefaultLabels(context, dataLabels, numeric = false, hasTooltip = false) {
    return newDrawDefaultLabels(context, downgradeToOldLabels(dataLabels), numeric, hasTooltip);
}
export function getDataLabelLayoutOptions(chartType) {
    return newGetDataLabelLayoutOptions(chartType);
}
export function getNumberOfLabelsToRender(viewportWidth, labelDensity, minimumLabelsToRender, estimatedLabelWidth) {
    return newGetNumberOfLabelsToRender(viewportWidth, labelDensity, minimumLabelsToRender, estimatedLabelWidth);
}
